source/manual/architecture-deep-dive.html.md

```diff
@@ -0,0 +1,462 @@
+
+#### Routing on the CDN
+
+As well as for caching, Varnish is used for the redirection from `gov.uk` to
+`www.gov.uk`, which is configured in Varnish Configuration Language (VCL) and
+uploaded directly to Fastly via [govuk-cdn-config].
+
+Other redirects that happen at the Fastly level include [bouncer]: a GOV.UK
+application responsible for redirecting traffic from old pre-GOV.UK websites.
+This is configured via [transition] and [transition-config].
+
+[bouncer]: https://github.com/alphagov/bouncer
+[govuk-cdn-config]: https://github.com/alphagov/govuk-cdn-config
+[transition]: https://github.com/alphagov/transition
+[transition-config]: https://github.com/alphagov/transition-config
+
+### Routing on GOV.UK
+
+#### Getting to the 'router' application
+
+![](images/load-balancer.png)
+
+Some requests make it through the CDN and cache layers to 'origin'. Origin is
+a stack of computers in the cloud - in this case, AWS - and its entry point is
+a load balancer.
+
+The load balancer knows based on the hostname which machine 'class' to route
+to. Different classes of machine run different sets of GOV.UK applications.
+How many machines are allocated to a class - and how big those machines are -
+is configured using [Terraform], in [govuk-aws]. What runs on each machine
+class is configured in govuk-puppet, a file and process management system we'll
+cover in more detail later.
+
+![](images/nginx-routing.png)
+
+External requests are routed to a 'cache' machine, which uses Varnish for its
+caching. If Varnish has the route response in its cache, it returns that,
+otherwise it forwards the request to [Nginx], a web server running on the
+machine. Nginx is configured with govuk-puppet.
+
+Nginx proxies most requests to [router], which is a GOV.UK maintained
+application running on the cache machines. However, Nginx does proxy some
+routes directly to other apps, such as image URLs routing to asset-manager.
+
+[govuk-aws]: https://github.com/alphagov/govuk-aws
+[Nginx]: https://www.nginx.com/
+[router]: https://github.com/alphagov/router
+[Terraform]: https://www.terraform.io/
+
+#### Routing via 'router'
+
+'router' is a reverse proxy app written in Go. It stores all known routes
+in memory using a [prefix trie]. The routes are loaded into the trie from
+[redis] at startup. Matched routes are proxied through to the relevant
+rendering app based on the `backend_id` of the route, which is derived from
+the `rendering_app` field in the corresponding content item in the content
+store (which we'll cover later). For example, if the route has a `backend_id`
+of `frontend`, it will forward the request to the [frontend] application.
+
+When routes are deleted, they're marked as `gone` in router, and router
+returns a 410 Gone response. When routes are changed (which can happen when a
+document's title is changed, affecting its slug), it is marked as `redirect`
+in router, serving a 301 Moved Permanently response. It's worth noting the
```

Comment:
I'm not sure about this bit. The Publishing API handles setting up redirects automatically when content is moved around, not the Router. It just listens to what the Content Store says.

Reply:
Yes, when a publishing app publishes a new edition with a new slug, Publishing API is aware the route for the content has changed and makes an [additional request to content store to set up a redirect](https://github.com/alphagov/publishing-api/blob/1b8e328540ab96759807787ab87dcb33bbcc72e3/spec/integration/put_content/content_with_a_previously_published_edition_spec.rb#L99-L132).

Content store [registers the redirect with router-api](https://github.com/alphagov/content-store/blob/08f02f990e621c9d2fd473e12a70a6805ddd8dcb/app/models/route_set.rb#L61-L64), which [stores the redirect in MongoDB](https://github.com/alphagov/router-api/blob/96e2838cf68158be7a3f4375e68b9a8c93991b68/app/models/route.rb). This route gets loaded into Router with a `redirect` handler instead of a `backend_id`, I think.

Does that sound right? I haven't been overly explicit about how it all works because I've tried introducing each concept gradually, but I don't think I've written anything here that contradicts that? 🤔

Reply:
Yep, that sounds right.

However, given this paragraph and the one prior to it are just about the Router, and the things it does. I would read the bit below as implying that this is a function of the Router, it watches for changes in routes, and sets up redirects.

> When routes are changed (which can happen when a document's title is changed, affecting its slug), it is marked as `redirect` in router, serving a 301 Moved Permanently response.

Given this section is about the router specifically, maybe it's best to bring this up when the Publishing API is mentioned instead?

Reply:
Have now rewritten this to be clearer 👍